Aidan/dockerfileless mode (#232)
latchbio/latch: docs/source/getting_started/authoring_your_workflow.md

# Authoring your Own Workflow

In this demonstration, we will examine a workflow which sorts and assembles COVID sequencing data.

> The document aims to be an extension to the quickstart to help you better understand the structure of a workflow and write your own.

Reviewer: ```suggestion
This document aims to be an extension to the Quickstart to help you better understand the structure of a workflow and write your own.
```

> This document aims to be an extension to the Quickstart to help you better understand the structure of a workflow and write your own.

Reviewer 2: add link to "Quickstart"

## 1: Initialize Workflow Directory
Bootstrap a new workflow directory by running `latch init` from the command line.

```shell-session

> $ latch init covid-wf -t subprocess

Reviewer 2: ```suggestion
$ latch init covid-wf --template subprocess
```

Created a latch workflow in `covid-wf`
Run
        $ latch register covid-wf
To register the workflow with console.latch.bio.
```

File Tree:
```shell-session
covid-wf
├── LICENSE
├── README.md
├── bowtie2
│   ├── bowtie2
│   └── ...
├── reference
│   ├── wuhan.1.bt2
│   └── ...
├── system-requirements.txt
├── version
└── wf
    ├── __init__.py
    ├── assemble.py
    └── sort.py
```

Once your boilerplate workflow has been created successfully, you should see a folder called `covid-wf`.

## 2. Writing your First Task

> A task is a Python function that takes in inputs and returns outputs to the Latch platform or to another task. In the `assemble.py` file inside the `covid-wf/wf` directory, we have a task which ingests two sequencing reads and outputs an assembled SAM file.

Reviewer 2: "we have" is silly tone

> ### How to work with LatchFiles and LatchDirs

Reviewer 2: ```suggestion
### Working with `LatchFile`s and `LatchDir`s
```

> `LatchFile` and `LatchDir` are types built into the Latch SDK which allow users to use files and directories on Latch as inputs and outputs to workflows. They point to remote file locations in our user interface on [Latch Console](https://console.latch.bio/data) and implement the necessary operations to ensure data is pulled into the task environment and saved back into your data tab.

Reviewer 2: add :class: references to `LatchFile` and `LatchDir`

Reviewer 2: "use files and directories on Latch" is confusing
suggestion: "use files and directories stored remotely on Latch"

Reviewer 2: would prefer "the Latch Console" over "Latch Console"

Reviewer 2: "pulled into" is silly
suggestion: "available in"

"saved back" is silly
suggestion: "outputs are uploaded to your Latch account"

> Any time you read or resolve the path of a file or directory in a task, the data will be downloaded locally and will be accessible within the task. For example the line `local_file = Path(read1)` in the snippet below will download the file from Latch and return a path to the local file.

Reviewer 2: "Any time" -> "The first time"
"downloaded locally" -> "downloaded"
"path to the local file" -> "path to the local copy"

```python
from latch.types import LatchFile, LatchOutputDir
...

@small_task
def assembly_task(
    read1: LatchFile, read2: LatchFile, output_directory: LatchOutputDir
) -> LatchFile:

    local_file = Path(read1)
    ...
```

> Sometimes, a python task does not directly access a file/directory, but uses it implicitly; for example, through a subprocessed shell command. We can ensure that the file's data is present by either calling `Path` on the file or directory, or by accessing the `local_path` attribute of the LatchFile or LatchDir. This will download the data into the task environment and return back a path to the local file or directory.

Reviewer 2: "but uses it implicitly" -> "but uses its path"
instead of ";", split into two sentences

"We can ensure that the file's data is present by either calling `Path` on the file or directory, or by accessing the `local_path` attribute of the LatchFile or LatchDir...." weird, especially given that the last paragraph already explained how to download shit

```python
bowtie2_cmd = [
    "bowtie2/bowtie2",
    "--local",
    "--very-sensitive-local",
    "-x",
    "wuhan",
    "-1",
    read1.local_path,
    "-2",
    read2.local_path,
    "-S",
    str(sam_file),
]
```

> A file or directory can be outputted to the Latch platform as below:

Reviewer 2: very confusing. explain that returning a `LatchFile`/`LatchDir` from the task/workflow will upload it to the specified remote path

```python
    local_file = Path("/root/covid_assembly.sam")
    return LatchFile(local_file, "latch:///covid_assembly.sam")
```

> Here, `LatchFile()` takes as input two values: the first being the local filepath and the second being the desired remote file location on Latch. Every file on Latch must have the prefix `latch:///`.

Reviewer 2: suggestion: `LatchFile(...)`

"the first being" -> "the first is"
"the second being" -> "the second is"
"desired" -> "target"

the prefix thing is strictly speaking wrong because we have other prefixes (in new lytekit) but doesn't need to be changed yet
"every file on Latch must have" -> "Remote paths on Latch start with"

## 1: Initialize Workflow Directory
Bootstrap a new workflow directory by running `latch init` from the command line.

> ```shell-session
> latch init covid-wf -t subprocess
> ```
> Output:
> ```shell-session
> Created a latch workflow in `covid-wf`
> Run
>         $ latch register covid-wf
> To register the workflow with console.latch.bio.

Reviewer: ```suggestion
```shell-session
$ latch init covid-wf -t subprocess
Created a latch workflow in `covid-wf`
Run
        $ latch register covid-wf
To register the workflow with console.latch.bio.
```
```

```
File Tree:
```shell-session
covid-wf
├── LICENSE
├── README.md
├── bowtie2
│   ├── bowtie2
│   └── ...
├── reference
│   ├── wuhan.1.bt2
│   └── ...
├── system-requirements.txt
├── version
└── wf
    ├── __init__.py
    ├── assemble.py
    └── sort.py
```

Once your boilerplate workflow has been created successfully, you should see a folder called `covid-wf`.

## 2. Writing your First Task

> A task is a Python function that takes in inputs and returns outputs to the latch platform or to another task. In the `assemble.py` file inside the `covid-wf/wf` directory, we have a task which ingests two sequencing reads and outputs an assembled SAM file. Take a look within the file.

Reviewer: ```suggestion
A task is a Python function that takes in inputs and returns outputs to the Latch platform or to another task. In the `assemble.py` file inside the `covid-wf/wf` directory, we have a task which ingests two sequencing reads and outputs an assembled SAM file.
```

### How to work with LatchFiles and LatchDirs

> `LatchFile` and `LatchDir` are types built into the Latch SDK which allow users to use files and directories on Latch as inputs and outputs to workflows. They point to remote file locations in our user interfact on [Latch Console](https://console.latch.bio/data) and implement the necessary operations to ensure data is pulled into the task environment and saved back into your data tab.

Reviewer: ```suggestion
`LatchFile` and `LatchDir` are types built into the Latch SDK which allow users to use files and directories on Latch as inputs and outputs to workflows. They point to remote file locations in our user interface on [Latch Console](https://console.latch.bio/data) and implement the necessary operations to ensure data is pulled into the task environment and saved back into your data tab.
```

> Anytime you read from a file or directory in a task, for example running the following code in a task

Reviewer: ```suggestion
Any time you read from a file or directory in a task, for example running the following code in a task
```

```python
from latch.types import LatchFile, LatchOutputDir
...

@small_task
def assembly_task(
    read1: LatchFile, read2: LatchFile, output_directory: LatchOutputDir
) -> LatchFile:

    local_file = Path(read1)
    ...
```

> The data will be downloaded locally and will be able to be accessed inside the task.

Reviewer: ```suggestion
the data will be downloaded locally and will be accessible inside the task.
```

since this is a continuation of the previous sentence

> Sometimes, our python task does not access a file, but the file data still needs to exist locally. For example, when running a subprocess with a file as an argument. We can get around having to call Path on the file or directory by accessing the `local_path` attribute of the LatchFile or LatchDir. This will download the data into the task and return back a path to the local file or directory.

Reviewer: ```suggestion
Sometimes, a python task does not directly access a file/directory, but uses it implicitly; for example, through a subprocessed shell command. We can ensure that the file's data is present by either calling `Path` on the file or directory, or by accessing the `local_path` attribute of the LatchFile or LatchDir. This will download the data into the task environment and return back a path to the local file or directory.
```

    
```python
bowtie2_cmd = [
    "bowtie2/bowtie2",
    "--local",
    "--very-sensitive-local",
    "-x",
    "wuhan",
    "-1",
    read1.local_path,
    "-2",
    read2.local_path,
    "-S",
    str(sam_file),
]
```

> To output a file or directory to the latch platform:

Reviewer: ```suggestion
A file or directory can be outputted to the Latch platform as below:
```

```python
    local_file = Path("/root/covid_assembly.sam")
    return LatchFile(local_file, "latch:///covid_assembly.sam")
```

> Here, LatchFile takes two values: the first being your local filepath and the second being the remote file location on Latch. Every file on Latch must have the prefix `latch:///`.

Reviewer: ```suggestion
Here, `LatchFile()` takes as input two values: the first being the local filepath and the second being the desired remote file location on Latch. Every file on Latch must have the prefix `latch:///`.
```

## 4. Manage installation for third-party dependencies

> Under the hood, Latch uses Dockerfiles for dependency management, which allows a user to define the computing environment that their task will execute in. In most cases, there is no need for a user to write a Dockerfile, as the Latch SDK will automatically generate one.

Reviewer 2: - mention the preferred option first
- more factual tone
```suggestion
Internally, Latch uses `Dockerfile`s for dependency management, which are automatically generated from environment files or handwritten.
```

> We need `samtools` and `autoconf`. We can install these dependencies by adding them to the `system-requirements.txt` file, as shown [here](../basics/defining_environment.md#system-debian-packages). We will create and add the following lines to the file:

Reviewer 2: grug need samtools. grug can install dependency by adding to system-requirements. grug will create and add the following lines

Reviewer 2: ```suggestion
Install `samtools` and `autoconf` by [adding them to the dependency list in `system-requirements.txt`:](../basics/defining_environment.md#system-debian-packages)
```

```text
samtools
autoconf
```

> Additionally, we need to set an environment variable for bowtie2 to work. As shown [here](../basics/defining_environment.md#environment-variables), we can set environment variables by adding them to the `environment` file. We will create and add the following line to the file:

Reviewer 2: ```suggestion
Set an environment variable for bowtie2 using [an environment file](../basics/defining_environment.md#environment-variables):
```

```text
BOWTIE2_INDEXES=reference
```

> After adding these files to the workflow directory, we can run `latch dockerfile .` to examine the auto generated Dockerfile.

Reviewer 2: ```suggestion
Now run `latch dockerfile .` to examine the auto generated Dockerfile:
```

```Dockerfile
# latch base image + dependencies for latch SDK --- removing these will break the workflow
from 812206152185.dkr.ecr.us-west-2.amazonaws.com/latch-base:ace9-main
run pip install latch==2.12.1
run mkdir /opt/latch

# install system requirements
copy system-requirements.txt /opt/latch/system-requirements.txt
run apt-get update --yes && xargs apt-get install --yes </opt/latch/system-requirements.txt

# copy all code from package (use .dockerignore to skip files)
copy . /root/

# set environment variables
env BOWTIE2_INDEXES=reference

# latch internal tagging system + expected root directory --- changing these lines will break the workflow
arg tag
env FLYTE_INTERNAL_IMAGE $tag
workdir /root
```

> For more information on how to configure your environment for other use cases, such as R or conda, see the [defining environment](../basics/defining_environment.md) section.

Reviewer 2: ```suggestion
See the [Workflow Environment] page(../basics/defining_environment.md) for further information configuring the environment for other use dependencies such as R or Conda.
```

## 4. Manage installation for third-party dependencies

> Under the hood, latch uses Dockerfiles for dependencies management, which allow you to define the computing environment that your task will execute in. In most cases, there is no need for a user to write a Dockerfile, as the Latch SDK will automatically generate one for you.

Reviewer: ```suggestion
Under the hood, Latch uses Dockerfiles for dependency management, which allows a user to define the computing environment that their task will execute in. In most cases, there is no need for a user to write a Dockerfile, as the Latch SDK will automatically generate one.
```

> We need `samtools` and `autoconf`. We can install these dependencies by adding them to the `system-requirements.txt` file, as shown [here](../basics/defining_environment.md#system-debian-packages). We will create and add the following lines to the file:

Reviewer: ```suggestion
We need `samtools` and `autoconf`. We can install these dependencies by adding them to the `system-requirements.txt` file, as shown [here](../basics/defining_environment.md#system-debian-packages). We will create and add the following lines to the file:
```

does the link exist?

> ```

Reviewer: ```suggestion
```plaintext
```

samtools
autoconf
```

Additionally, we need to set an environment variable for bowtie2 to work. As shown [here](../basics/defining_environment.md#environment-variables), we can set environment variables by adding them to the `environment` file. We will create and add the following line to the file:

> ```

Reviewer: ```suggestion
```plaintext
```

BOWTIE2_INDEXES=reference
```

> After adding these files to the workflow directory, we can run `latch dockerfile .` to check out the auto generated Dockerfile.

Reviewer: ```suggestion
After adding these files to the workflow directory, we can run `latch dockerfile .` to examine the auto generated Dockerfile.
```

```Dockerfile
# latch base image + dependencies for latch SDK --- removing these will break the workflow
from 812206152185.dkr.ecr.us-west-2.amazonaws.com/latch-base:ace9-main
run pip install latch==2.12.1
run mkdir /opt/latch

# install system requirements
copy system-requirements.txt /opt/latch/system-requirements.txt
run apt-get update --yes && xargs apt-get install --yes </opt/latch/system-requirements.txt

# copy all code from package (use .dockerignore to skip files)
copy . /root/

# set environment variables
env BOWTIE2_INDEXES=reference

# latch internal tagging system + expected root directory --- changing these lines will break the workflow
arg tag
env FLYTE_INTERNAL_IMAGE $tag
workdir /root
```

> To configure your environment for other use cases, such as R or conda, visit the [defining environment](../basics/defining_environment.md) section.

Reviewer: ```suggestion
For more information on how to configure your environment for other use cases, such as R or conda, see the [defining environment](../basics/defining_environment.md) section.
```
